### DISCUSSION TITLE
A visualization problem occurs after NetworkX is imported into Mesa

### DISCUSSION THREAD

**Discussion by bruce881225** [mesa/mesa#1816] [Q&A]
Hello, everyone!
First of all, I want to thank you for the great work you've done for Mesa.
I am new to python and mesa, and I am trying to place the Agent on all nodes of a simple NetworkX digraph, with the colors and positions in the node attributes representing the Agent's initial positions and colors. However, the result of my visualization is different from the NetworkX digraph I created. How can I modify my code?
(I haven't decided what the Agent's behavior is yet, so I'll represent it with a color change.)
![1](https://github.com/projectmesa/mesa/assets/145688841/5783ee86-5a15-44dc-a39a-c455777fabbf)
![2](https://github.com/projectmesa/mesa/assets/145688841/9bf02a1c-3895-488e-8431-0b34f9ac28a1)
![3](https://github.com/projectmesa/mesa/assets/145688841/201c3781-ad94-4e99-99cf-fb460a4abe32)
![4](https://github.com/projectmesa/mesa/assets/145688841/f34da64b-6707-4bde-ab60-81dba218c502)
![5](https://github.com/projectmesa/mesa/assets/145688841/2b91892d-6f88-4293-a99c-b12eb1942749)
![6](https://github.com/projectmesa/mesa/assets/145688841/c2e81fa3-43a5-4901-80dd-db5508c7f7ac)
![7](https://github.com/projectmesa/mesa/assets/145688841/c5305525-187e-4546-9d45-cb19b32254f4)

**Comment by rht** [comment]
The frontend visualization is rendered using JavaScript via D3.js.

We have a new experimental Python-only frontend that uses NetworkX, but the positions are still hardcoded to be in a spring layout: https://github.com/projectmesa/mesa/blob/2be2f05ee6e5a68e0fcabd90f1f43c3ed22b4c21/mesa/experimental/jupyter_viz.py#L276-L284. To use this, you have to install from the latest main on Git. Tutorial on the new frontend can be found at https://mesa.readthedocs.io/en/stable/tutorials/visualization_tutorial.html.

This is an example of it: https://github.com/projectmesa/mesa-examples/pull/56 (not yet merged because we haven't released a version with the NetworkX frontend yet, but a release is expected by the end of the month).

To override the current network space drawer, you can instead write your own custom space draw. See e.g. https://github.com/projectmesa/mesa-examples/blob/main/examples/sugarscape_g1mt/app.py. But the API for the function signature has changed from `def space_drawer(viz):` to `def space_drawer(model, agent_portrayal):`.

This hasn't been properly documented because it's very new. You could also submit a pull request to modify the network drawer to detect that the nodes have pos attribute, and will use that instead of generating a spring layout.

**Reply by rht** [reply]
I didn't finish on elaborating this sentence:
> The frontend visualization is rendered using JavaScript via D3.js.

The JS frontend hardcodes the layout to a spring layout. But it's easier to use the new Python frontend to make custom change instead of modifying the JS frontend.
